File: Utilities/DatabaseUpdater/CSharpProjectScriptSource.vb
Imports System.Collections.Generic
Imports System.Text.RegularExpressions
Public Class CSharpProjectScriptSource
    Inherits ScriptSource
    Friend path As String
    Friend current As Integer
    Friend fileContent As String()
    Friend currentPath As String
    Friend depth As Integer = 0


    Sub New(ByVal path As String)
        Me.path = path
    End Sub

    Public Overrides Sub FillTables()
        currentPath = New IO.FileInfo(path).Directory.FullName + "\"
        fileContent = System.IO.File.ReadAllLines(path)
        current = 0
        While current < fileContent.Length
            If IsScriptDefinition(Line) Then
                Dim path As String = GetBitInQuotes()
                Select Case GetFirstBit(path)
                    Case "Change Scripts"
                        AddScript(ScriptType.UpdateScript, path)
                    Case "Create Scripts"
                        Select Case GetSecondBit(path)
                            Case "prc" : AddScript(ScriptType.StoredProcedure, path)
                            Case "udf" : AddScript(ScriptType.UserDefinedFunction, path)
                            Case "viw" : AddScript(ScriptType.View, path)
                            Case "trg" : AddScript(ScriptType.Trigger, path)
                        End Select
                    Case "Data Scripts"
                        AddScript(ScriptType.TableData, path)
                End Select
            End If
            MoveToNextLine()
        End While
    End Sub

    Public Shared Function GetFirstBit(ByVal path As String) As String
        Return path.Substring(0, path.IndexOf("\"))
    End Function

    Public Shared Function GetSecondBit(ByVal path As String) As String
        Dim indexOfFirstSlash As Integer = path.IndexOf("\")
        Return path.Substring(indexOfFirstSlash + 1, path.IndexOf("\", indexOfFirstSlash + 1))
    End Function


    Sub MoveToNextLine()
        current += 1
    End Sub
    ReadOnly Property Line() As String
        Get
            If current < fileContent.Length Then
                Return Me.fileContent(current).Trim
            Else
                Return ""
            End If
        End Get
    End Property
    Function GetBitInQuotes() As String
        Return Line.Substring(Line.IndexOf("""") + 1, Line.LastIndexOf("""") - Line.IndexOf("""") - 1).Replace("""""", """")
    End Function

    Shared Function IsScriptDefinition(ByVal line As String) As Boolean
        Try
            Return line.IndexOf("<None Include=""") > -1 And line.IndexOf(".sql"" />") > -1
        Catch ex As System.ArgumentOutOfRangeException
            Return False
        End Try

    End Function
End Class

File: Utilities/SqlScriptRunner/ScriptSources/CSharpProjectScriptSource.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SqlScriptRunner.ScriptSources
{

    internal class CSharpProjectScriptSource : ScriptSource
    {
        internal System.IO.FileInfo projectFile;
        internal int current;
        internal string[] fileContent;

        internal CSharpProjectScriptSource(string path)
        {
            this.projectFile = new System.IO.FileInfo(path);
            fileContent = System.IO.File.ReadAllLines(projectFile.FullName);
            current = 0;
            while (current < fileContent.Length)
            {
                if (IsScriptDefinition(Line))
                {
                    string relativePathToScript = GetBitInQuotes();

                    string absolutePathToScript = System.IO.Path.Combine(projectFile.DirectoryName, relativePathToScript);
                    string scriptType = null;
                    if (absolutePathToScript.IndexOf("\\" + ScriptSource.UpdateScriptDir  + "\\") > -1) {scriptType = ScriptType.UpdateScript ;}
                    if (absolutePathToScript.IndexOf("\\" + ScriptSource.CreateScriptDir  + "\\" + ScriptSource.ProcedureSubDir + "\\") > -1) { scriptType = ScriptType.StoredProcedure; }
                    if (absolutePathToScript.IndexOf("\\" + ScriptSource.CreateScriptDir  + "\\" + ScriptSource.FunctionSubDir + "\\") > -1) { scriptType = ScriptType.UserDefinedFunction; }
                    if (absolutePathToScript.IndexOf("\\" + ScriptSource.CreateScriptDir  + "\\" + ScriptSource.ViewSubDir + "\\") > -1) { scriptType = ScriptType.View; }
                    if (absolutePathToScript.IndexOf("\\" + ScriptSource.CreateScriptDir  + "\\" + ScriptSource.TriggerSubDir + "\\") > -1) { scriptType = ScriptType.Trigger; }
                    if (absolutePathToScript.IndexOf("\\" + ScriptSource.DataScriptDir    + "\\") > -1) { scriptType = ScriptType.TableData; }
					if (absolutePathToScript.IndexOf("\\" + ScriptSource.OnSyncScriptDir + "\\") > -1) { scriptType = ScriptType.OnSync; }
                  
                    AddScript(scriptType, absolutePathToScript);
                }
                MoveToNextLine();
            }
        }

        internal  static string GetFirstBit(string path)
        {
            return path.Substring(0, path.IndexOf("\\"));
        }

        internal  static string GetSecondBit(string path)
        {
            int indexOfFirstSlash = path.IndexOf("\\");
            return path.Substring(indexOfFirstSlash + 1, path.IndexOf("\\", indexOfFirstSlash + 1));
        }

        void MoveToNextLine()
        {
            current += 1;
        }

        string Line
        {
            get
            {
                if (current < fileContent.Length)
                {
                    return this.fileContent[current].Trim();
                }
                else
                {
                    return "";
                }
            }
        }

        string GetBitInQuotes()
        {
            return Line.Substring(Line.IndexOf("\"") + 1, Line.LastIndexOf("\"") - Line.IndexOf("\"") - 1).Replace("\"\"", "\"");
        }

        static bool IsScriptDefinition(string line)
        {
            try
            {
                return line.IndexOf("<None Include=\"") > -1 & line.IndexOf(".sql\" />") > -1;
            }
            catch (System.ArgumentOutOfRangeException ex)
            {
                return false;
            }
        }
    }
}
